app/controllers/admin/enterprise_images_controller.rb

```diff
@@ -0,0 +1,41 @@
+module Admin
```

Comment:
I think this should be an API endpoint, not an admin endpoint. This controller should be on /api. Actually, I think it should be on /api/enterprises folder.

Reply:
@luisramos0 Do you mean `/admin/api` or just `/api`? I never thought about the mix of admin and customer code in the API before. I guess it would make sense to have *one* `/api` for everything that is really consistent. Would that be its own engine later on?

Reply:
I checked just now, and it seems that some admin actions for products already use `/api`:

```bash
$ git grep /api app/assets/javascripts/admin/
app/assets/javascripts/admin/bulk_product_update.js.coffee:137:          url: "/api/products/" + product.id + "/soft_delete"
app/assets/javascripts/admin/bulk_product_update.js.coffee:152:              url: "/api/products/" + product.permalink_live + "/variants/" + variant.id + "/soft_delete"
app/assets/javascripts/admin/index_utils/services/resources.js.coffee:2:  LineItem = $resource '/api/orders/:order_number/line_items/:line_item_id.json',
app/assets/javascripts/admin/index_utils/services/spree_api_auth.js.coffee:6:      $http.get("/api/users/authorise_api?token=" + SpreeApiKey)
app/assets/javascripts/admin/products/services/product_image_service.js.coffee:11:      @imageUploader.url = "/api/product_images/#{product.id}"
app/assets/javascripts/admin/services/bulk_products.js.coffee:10:      url = "/api/products/bulk_products?page=::page::;per_page=20;#{queryString}"
app/assets/javascripts/admin/services/bulk_products.js.coffee:15:      $http.post("/api/products/" + product.id + "/clone").success (data) =>
app/assets/javascripts/admin/services/bulk_products.js.coffee:16:        dataFetcher("/api/products/" + data.id + "?template=bulk_show").then (newProduct) =>
app/assets/javascripts/admin/variant_overrides/controllers/variant_overrides_controller.js.coffee:46:    url = "/api/products/overridable?page=::page::;per_page=100"
```

Should we already be building actions within `/api` as much as possible? I think that would be ideal, to take more advantage of using Angular and having an API.

Reply:
I think /api is better at this stage.
In terms of engines, this is a good example where admin is not appropriate because this is called from the frontoffice. Anyway, this code would go into the "Entities" domain  (Entity Management/Enterprises).

My proposal (in line with Enrico's suggestion) is to have the API code inside each engine. So each engine will have it's own models, controllers, views, api and angular UI, including some rails views and it's html controllers as well as api controllers and angular code.

In terms of api endpoints and engines we can have two options:
- option 1 - engine name on the api endpoint - /entities/api/enterprises or /web/api/content/ or /shop/api/cart, entities, web and shop being the domains.
- option 2 - no engine name on the api endpoint - all engines make endpoints at /api like /api/enterprises /api/content /api/cart

imo, I think we should aim for the first option with domain name on api endpoints but I don't think we should go for it right now. So, I think /entities/api/enterprises is better but for now I'd put it as /api/enterprises.

I'll put this conversation on the domains thread.

Reply:
Okay, let's go with `/api/`. We still have the option between `/api/enterprise_image` and `/api/enterprise/:id/logo`. The latter seems better structured to me. To really follow Rails' RESTful approach, we would have one controller for the logo and one for the promo image.

:thinking: 

Looking at the current logic, we could create a new model `EnterpriseAttachment` which contains the logic that is currently in protected methods (load enterprise and delete attachment). With such a model, it could be better to have an API like `/api/enterprise/:id/attachment/:name`. That leaves us with only one controller that can handle logos and promo images and other future attachments.

What do you think @kristinalim? I'm not sure if we need the additional model. It could just help to keep the controller even smaller.

@luisramos0 Regarding the implications of the new engines approach: This is an interesting example. My approach would probably be to first think of the best API. If it makes sense to have domains like `shop` and `web` in the API then it would make sense to structure our code accordingly. But in the end, the API should be something stable that is independent from our internal code organisation. So our code structure should not define the API.

You mentioned that engines would have their own models. That was one of my questions about this structure. Wouldn't that lead to a lot of code duplication? I thought the idea of models is that they represent the data to manipulate which is independent of the perspective from which we look at them. It's probably best to continue this discussion on Discourse.

Reply:
@mkllnk Thank you! I prefer your first suggestion.

About using `/api/enterprise/:id/attachment/:name`, I have hesitations about abstracting the logo and promo image this way, which almost puts them on the same level of importance as each other and implying that they would always behave similarly.

I prefer think of them as two different resources which just happen to have the same type of storage and the same available actions. So, I'm more interested in your first suggestion, where we would have separate singular resource controllers for the logo and for the promo image, and then we can just have these inherit from an `EnterpriseAttachmentController` which would have most of the shared logic.